File: peakelement.cpp
//TC O(logn)
//SC O(1)
//here first we will check if it is the peak or not if ti is return it & taking care of the edge case
//if it is not peak we will move to the inc side since that will give us the peak element
class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        int n = nums.size();
        int s =0;
        int e = n-1;
        while(s<=e){
            int mid = s+(e-s)/2;
            if((mid==0 || nums[mid]>nums[mid-1]) && (mid ==e || nums[mid]>nums[mid+1])) return mid;
            else if(nums[mid]<nums[mid+1]) s =mid+1;
            else  e = mid-1;
        }
        return -1;
    }
};

File: PeakElement.cs
// Time Complexity : O(lon n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :Yes
// performed binary search, find mid and check where is the greater element
// then discard the other half
public class Solution {
    public int FindPeakElement(int[] nums) {
        if(nums == null || nums.Length ==0)
        {
            return -1;
        }
        int l = 0;
        int n = nums.Length;
        int h = n - 1;
        while (l <= h)
        {
            int mid = l +(h-l)/2;
            // check if mid is peak element
            if((mid == l || nums[mid] > nums[mid-1]) && (mid==h || nums[mid] > nums[mid+1] ))
            {
                return mid;
            }
            // if right element is greater discard left half
            else if(nums[mid+1] > nums [mid])
            {
                l= mid+1;         
            }
            else // discard right half
            {
                h= mid-1;
            }
        }
        return 434;
    }
}